addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/xml/InputWithPresetControlXML.java

```diff
@@ -0,0 +1,123 @@
+
+    private PresetInfoStateListener observer;
+
+    /**
+     * Create a InputWithPlayControl object for altering menu positions and requesting current menu information as well
+     * as controlling the playback and choosing a preset item.
+     *
+     * @param inputID The input ID like USB or NET_RADIO.
+     * @param com The Yamaha communication object to send http requests.
+     */
+    public InputWithPresetControlXML(String inputID, AbstractConnection com, PresetInfoStateListener observer) {
+        this.inputID = inputID;
+        this.comReference = new WeakReference<AbstractConnection>(com);
+        this.observer = observer;
+    }
+
+    /**
+     * Wraps the XML message with the inputID tags. Example with inputID=NET_RADIO:
+     * <NETRADIO>message</NETRADIO>.
+     *
+     * @param message XML message
+     * @return
+     */
+    protected String wrInput(String message) {
+        return "<" + inputID + ">" + message + "</" + inputID + ">";
+    }
+
+    /**
+     * Updates the preset information
+     *
+     * @throws Exception
+     */
+    public void update() throws IOException, ReceivedMessageParseException {
+        if (observer == null) {
+            return;
+        }
+
+        AbstractConnection com = comReference.get();
+        String response = com.sendReceive(wrInput("<Play_Control>GetParam</Play_Control>"));
+        Document doc = XMLUtils.xml(response);
+        if (doc.getFirstChild() == null) {
+            throw new ReceivedMessageParseException("<Play_Control>GetParam failed: " + response);
+        }
+
+        PresetInfoState msg = new PresetInfoState();
+
+        Node playbackInfoNode = XMLUtils.getNode(doc.getFirstChild(), "Play_Control/Preset/Preset_Sel");
+        msg.presetChannel = playbackInfoNode != null ? Integer.valueOf(playbackInfoNode.getTextContent()) : -1;
+
+        // Set preset channel names, obtained from this xpath:
+        // NET_RADIO/Play_Control/Preset/Preset_Sel_Item/Item_1/Title
+        playbackInfoNode = XMLUtils.getNode(doc.getFirstChild(), "Play_Control/Preset/Preset_Sel_Item");
+        if (playbackInfoNode != null) {
+            for (int i = 1; i <= PRESET_CHANNELS; ++i) {
+                Node itemNode = XMLUtils.getNode(playbackInfoNode, "Item_" + String.valueOf(i) + "/Title");
+                String v = itemNode != null ? itemNode.getTextContent() : "Item_" + String.valueOf(i);
+                if (!v.equals(msg.presetChannelNames[i - 1])) {
+                    msg.presetChannelNamesChanged = true;
+                    msg.presetChannelNames[i - 1] = v;
+                }
+            }
+        }
+
+        observer.presetInfoUpdated(msg);
+    }
+
+    /**
+     * Select a preset channel.
+     *
+     * @param number The preset position [1,40]
+     * @throws Exception
+     */
+    public void selectItemByPresetNumber(int presetChannel) throws IOException, ReceivedMessageParseException {
+        comReference.get().send(wrInput(
```

Comment:
I thought the same while adding Spotify support in another branch and refactored this - please see my earlier comment. Now please advise suggest how to go about this @martinvw :)